### ISSUE TITLE
Dependency confusion

### ISSUE THREAD

**Issue by matt-noonan** [commercialhaskell/stack#5488]
### General summary/comments

Suppose that an `extra-dep` is given for package`my-private-package`, pointing to a specific commit in a specific git repo. I had expected that this repo/commit will *always* be used to provide `my-private-package`. However, when the version of `my-private-package` in the repo matches the version of the same-named package in the LTS, then the LTS version appears to be used.

If the private package's version is greater *or* less than the LTS package, the private package is used as expected. The LTS version only seems to be preferred when the version matches on-the-node. Adding or removing version bounds in the cabal file did not seem to effect this behavior.

Because of this, stack projects that have dependencies living in private git repos are vulnerable to a dependency confusion attack: https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610

### Steps to reproduce

1. Make a new project `X` with the same package name *and version* as something that already exists in the LTS you want to use. You can use the default project structure from `stack new X`, so you'll get a `Lib` module with a `someFunc` export. 
2. Push the project into a git repo and note the repo's URL and the commit hash.
3. Make a new project `Y`, add `X` as a dependency in the cabal file, and add an extra-dep in `stack.yaml` that points to the git repo/commit from step 2. Use this for the `Main` module:
```
module Main where
import Lib (someFunc)
main :: IO ()
main = someFunc
```
4. Attempt to build `Y`

### Expected

The version of `X` from the git repo should be used to build `Y`.

### Actual

The version of `X` from the LTS is used instead.

### Stack version

```
$ stack --version
Version 2.5.1, Git revision d6ab861544918185236cf826cb2028abb266d6d5 x86_64 hpack-0.33.0
```

### Method of installation

* Official binary, downloaded from stackage.org or fpcomplete's package repository

**Comment by matt-noonan** [comment]
Repeating the same steps with `astro`, version `0.4.2.1`, LTS 17.2 failed to reproduce. Now I'm even more puzzled.

**Comment by matt-noonan** [comment]
More data points: I could reproduce for `binary`, not for `control-bool`. Could it have to do with which libraries are boot libs? If that were the case, it might not be a very risky bug after all, though it is still a bit weird.

**Comment by qrilka** [comment]
@matt-noonan it's highly probable that boot libraries could show different results - they are handled differently. As for packages from snapshots - my short look into [`fillProjectWanted`](https://github.com/commercialhaskell/stack/blob/master/src/Stack/Config.hs#L646) didn't show clear ways to have that possible.
I'll try to check the logic around boot libraries later.

**Comment by chrisdone** [comment]
Probably related [Mismatched global packages](https://www.snoyman.com/blog/2019/01/mismatched-global-packages/).

**Comment by qrilka** [comment]
@matt-noonan could you share your particular combination of packages with a reproduction? I couldn't reproduce the original scenario and without a reproduction reading through the code didn't bring any hints about what could be a cause of such a behavior...

**Comment by qrilka** [comment]
Ping @matt-noonan

**Comment by matt-noonan** [comment]
Sorry for the delay. I've pushed up the main repo here: [https://github.com/matt-noonan/confusion](https://github.com/matt-noonan/confusion). It relies on the fake `array` package, at various versions, found here: [https://github.com/matt-noonan/fake-array-test](https://github.com/matt-noonan/fake-array-test)

To reproduce:

1. Check out the `confusion` repo and do a `stack build`. It should succeed. 
2. Now change the `stack.yaml` file by commenting out the commit for `array-0.5.3.9` and uncommenting the commit for `array-0.5.4.0`. Do a `stack build`. It should fail.
3. Finally, comment out the commit for `array-0.5.4.0` and uncomment the commit for `array-0.5.4.1`. Do a `stack build`. It should succeed.

**Comment by qrilka** [comment]
Thanks a lot

**Comment by qrilka** [comment]
Confirmed, need to dig deeper into what could be a reason for this behaviour

**Comment by qrilka** [comment]
It looks like the reason behind this is a bit special status of boot libraries (and extra package DBs as well BTW) - they are not managed by Stack, come with GHC thus they have a bit special handling. In the current implementation packages coming with GHC and the ones supplied in extra DBs can't be overridden by both local and snapshot packages if version are the same.
I don't see an easy fix for this and the problem doesn't appear to cause much of a trouble so I don't plan to resolve this. But I'll be glad to help answering any questions on the topic if this would help somebody else to work on this.

**Comment by fizbin** [comment]
Maybe this could be turned into a "document the behavior around..." ticket, and closed once appropriate documentation is added to make it clear that this dependency confusion route applies only when your pinned-to-git-commit package collides with the name of a package that comes with GHC.

**Comment by qrilka** [comment]
Yes, it probably makes sense to document this behavior if there are no immediate plans to fix it. But I'm still inclined to name it rather a bug than a feature :)

**Comment by mpilgrem** [comment]
Closing, as documentation now updated.
